thin-replica-server/include/thin-replica-server/thin_replica_impl.hpp

```diff
@@ -933,72 +933,84 @@ class ThinReplicaImpl {
-      ConcordAssertLE(global_eg_id_to_read, live_updates->newestEventGroupId());
-      return;
-    }
 
-    // Gap:
-    // If the first live update is not the follow-up to the last read event group from
-    // KVB then we need to fill the gap. Let's read from KVB starting at end + 1
-    // up to updates that are part of the live updates already. Thereby, we
-    // create an overlap between what we read from KVB and what is currently in
-    // the live updates.
-    if (live_updates->oldestEventGroupId() > global_eg_id_to_read) {
-      ConcordAssertGT(kvb_filter->newestExternalEventGroupId(), end + 1);
-      start = end + 1;
-      end = kvb_filter->newestExternalEventGroupId();
+      // fill the gap
+      LOG_INFO(logger_,
+               "Sync filling gap [" << next_global_eg_id_to_read << ", " << live_updates->oldestEventGroupId() << "]");
+      while (live_updates->oldestEventGroupId() > next_global_eg_id_to_read) {
+        // gap exists between next_global_eg_id_to_read and live_updates->oldestEventGroupId()
+        // Let's read from data table directly, filter and send updates, until next_global_eg_id_to_read equals
+        // live_updates->oldestEventGroupId()
+        auto event_group = kvb_filter->getEventGroup(next_global_eg_id_to_read);
+        if (event_group.events.empty()) {
+          std::stringstream msg;
+          msg << "EventGroup empty/doesn't exist for global event group " << next_global_eg_id_to_read;
+          throw kvbc::KvbReadError(msg.str());
+        }
+        auto eg_update_from_storage = kvbc::EgUpdate{event_group.id, std::move(event_group)};
+        auto filtered_eg_update = kvb_filter->filterEventGroupUpdate(eg_update_from_storage);
+        if (!filtered_eg_update) {
+          next_global_eg_id_to_read++;
+          continue;
+        }
+        // Overwrite event group ID in the filtered update to external event group ID
+        // We don't want to expose the global event group ID to the client
+        filtered_eg_update.value().event_group_id = ext_eg_id;
 
-      LOG_INFO(logger_, "Sync filling gap [" << start << ", " << end << "]");
-      readAndSendEventGroups<ServerContextT, ServerWriterT, DataT>(logger_, context, stream, start, end, kvb_filter);
-    }
+        kvb_filter->setNextExternalEgIdToRead(++ext_eg_id);
+        kvb_filter->setLastGlobalEgIdReadForClient(next_global_eg_id_to_read++);
 
-    last_global_eg_id_read = kvb_filter->getLastGlobalEgIdReadForClient();
-    // Overlap:
-    // If we read updates from KVB that were added to the live updates already
-    // then we just need to drop the overlap and return
-    ConcordAssert(live_updates->oldestEventGroupId() <= last_global_eg_id_read);
-    SubEventGroupUpdate update;
-    do {
-      live_updates->PopEventGroup(update);
-      LOG_INFO(logger_, "Sync dropping " << update.event_group_id);
-    } while (update.event_group_id < last_global_eg_id_read);
+        // send update
```

Comment:
Given the clear name of the following functions, I believe this comment adds no value.